src/metabase/driver/h2.clj

```diff
@@ -400,29 +398,17 @@
   [driver [_ field]]
   [:log10 (sql.qp/->honeysql driver field)])
 
-(defn- literal-text-value?
-  [[_ value {base-type :base_type effective-type :effective_type} :as clause]]
-  (and (mbql.u/is-clause? :value clause)
-       (string? value)
-       (isa? (or effective-type base-type)
-             :type/Text)))
-
-(defmethod sql.qp/->honeysql [:h2 :expression]
-  [driver [_ expression-name {::add/keys [source-table]} :as clause]]
-  (let [expression-definition (mbql.u/expression-with-name sql.qp/*inner-query* expression-name)]
-    (if (and (not= source-table ::add/source)
-             (literal-text-value? expression-definition))
-      ;; A literal text value gets compiled to a parameter placeholder like "?". H2 attempts to compile the prepared
-      ;; statement immediately, presumably before the types of the params are known, and sometimes raises an "Unknown
-      ;; data type" error if it can't deduce the type. The recommended workaround is to insert an explicit CAST. We
-      ;; only need to do this for string literals, since numbers and booleans get inlined directly.
-      ;;
-      ;; https://linear.app/metabase/issue/QUE-726/
-      ;; https://github.com/h2database/h2database/issues/1383
-      (->> (sql.qp/->honeysql driver expression-definition)
-           (h2x/cast :text))
-      ((get-method sql.qp/->honeysql [:sql-jdbc :expression])
-       driver clause))))
+(defmethod sql.qp/->honeysql [:h2 ::sql.qp/expression-literal-text-value]
+  [driver [_ value]]
+  ;; A literal text value gets compiled to a parameter placeholder like "?". H2 attempts to compile the prepared
+  ;; statement immediately, presumably before the types of the params are known, and sometimes raises an "Unknown
+  ;; data type" error if it can't deduce the type. The recommended workaround is to insert an explicit CAST. We
+  ;; only need to do this for string literals, since numbers and booleans get inlined directly.
```

Comment:
The last sentence is a bit confusing, because at this point we wouldn't be able to do anything about numbers or booleans anyway. The whole method is about handling strings, right?

Reply:
Good point. This part of the comment should probably be lifted into the `:sql` method to explain why it is restricted to string literals.